### DISCUSSION TITLE
Localization inherit and extend are not work well

### DISCUSSION THREAD

**Discussion by tian-jie** [abpframework/abp#19610] [General] [answered]
I am trying to use Abp Localization functions. Here is the design:
### base Localization module:
``` csharp
[DependsOn(typeof(AbpLocalizationModule))]
public class LocalizationBaseModule : AbpModule
{
    public override void ConfigureServices(ServiceConfigurationContext context)
    {
        Configure<AbpVirtualFileSystemOptions>(options =>
        {
            options.FileSets.AddEmbedded<LocalizationBaseModule>("Test.LocalizationBase");
        });

        Configure<AbpLocalizationOptions>(options =>
        {
            //Define a new localization resource
            options.Resources
                .Add<LocalizationBaseResource>("en-US")
                .AddVirtualJson("/Resources/Localization");
        });
    }
}
```
### extended Localization module:
``` csharp
[DependsOn(typeof(LocalizationBaseModule))]
public class MyLocalizationModule : AbpModule
{
    public override void ConfigureServices(ServiceConfigurationContext context)
    {
        Configure<AbpVirtualFileSystemOptions>(options =>
        {
            options.FileSets.AddEmbedded<MyLocalizationModule>("Test.MyLocalization");
        });

        Configure<AbpLocalizationOptions>(options =>
        {
            //Define a new localization resource
            options.Resources
                .Add<MyLocalizationResource>("en-US")
                .AddVirtualJson("/Resources/Localization")
                .AddBaseTypes(typeof(LocalizationBaseResource));
        });
    }
}
```

I tried once I use following in MyLocalizationModule, it will  **NOT** load localization resources from LocalizationBaseResource. if I remove it, it will not load resource from MyLocalizationResource.
``` csharp
options.FileSets.AddEmbedded<MyLocalizationModule>("Test.MyLocalization")
```

What is the best practice to using Abp.Localization in this case? My requirement is:
1. 2 different solutions
2. MyLocalizationResource will inherit resources from LocalizationBaseResource, that means, if LocalizationBaseResource has a resource key like "GetStart" and has localization in "en-US" to "Get start", and MyLocalizationResource does not have such key, when I instance IStringLocalizer<MyLocalizationResource>, it will get the translate "Get start"
3. MyLocalizationResource have ability to extend new keys which LocalizationBaseResource doesn't contain.
4. MyLocalizationResource have ability to replace keys which they both contain.

**Comment by tian-jie** [comment]
in addition, I am using abp framework as a wpf application.

**Comment by maliming** [comment, marked as answer]
hi

We have a related article.

https://community.abp.io/posts/how-to-override-localization-strings-of-depending-modules-ba1oy03l

**Reply by tian-jie** [reply]
Dear Maliming, 

Thanks for the reply!

I went through the code, I see you added embeded without parameter.
options.FileSets.AddEmbedded<ExtendLocalizationResourceDomainSharedModule>();

and you replaced embedded resource with physical file in your disk ( I believe this is the built out file which contains the resource)
options.FileSets.ReplaceEmbeddedByPhysical<ExtendLocalizationResourceDomainModule>(Path.Combine(hostingEnvironment.ContentRootPath, $"..{Path.DirectorySeparatorChar}ExtendLocalizationResource.Domain"));


in my testing, here is the point:
if I use options.FileSets.AddEmbedded<TestWpfModule>("TestWpf"); with namespace parameter, it works for the resource in this project, but resources in BaseType does not work.
If I use options.FileSets.AddEmbedded<TestWpfModule>(); without namespace parameter or not add this line, it works fine for BaseType resources, but resource in current project is not work.

Here is my test code:
base type class:
``` csharp
namespace TestWpf.Localization;

[DependsOn(typeof(AbpLocalizationModule))]
public class LocalizationModule : AbpModule
{
    public override void ConfigureServices(ServiceConfigurationContext context)
    {
        Configure<AbpVirtualFileSystemOptions>(options =>
        {
            // "YourRootNameSpace" is the root namespace of your project. It can be empty if your root namespace is empty.
            options.FileSets.AddEmbedded<LocalizationModule>("TestWpf.Localization");
        });
        Configure<AbpLocalizationOptions>(options =>
        {
            options.Languages.Add(new LanguageInfo("en-US", "English", "famfamfam-flag-england"));
            options.Languages.Add(new LanguageInfo("zh-Hans", "English", "famfamfam-flag-england"));

            options.Resources
                .Add<LocalizationResources>()
                .AddVirtualJson("/Resources/Localization")
                .DefaultCultureName = "zh-Hans";

            options.DefaultResourceType = typeof(LocalizationResources);

        });
    }
}
```

inherited class
``` csharp
namespace TestWpf;

[DependsOn(typeof(AbpAutofacModule))]
[DependsOn(typeof(Localization.LocalizationModule))]
public class TestWpfModule : AbpModule
{
    public override void ConfigureServices(ServiceConfigurationContext context)
    {
        context.Services.AddSingleton<MainWindow>();

        Configure<AbpVirtualFileSystemOptions>(options =>
        {
            // "YourRootNameSpace" is the root namespace of your project. It can be empty if your root namespace is empty.
            options.FileSets.AddEmbedded<TestWpfModule>("TestWpf");
        });
        Configure<AbpLocalizationOptions>(options =>
        {
            options.Languages.Add(new LanguageInfo("en-US", "English", "famfamfam-flag-england"));
            options.Languages.Add(new LanguageInfo("zh-Hans", "English", "famfamfam-flag-england"));

            options.Resources
                .Get<LocalizationResources>()
                .AddVirtualJson("/Resources/Localization/")
                .DefaultCultureName = "zh-Hans";
            options.DefaultResourceType = typeof(LocalizationResources);
        });

        context.Services.AddLocalization();
    }
}
```

here is the running screenshot, which means: 
**Hello US** is from current project(UI)
**GoodbyeWorld** is from base type module, which is **not** localized, this is original language key. this is wrong. I want it be localized based on base type resource.
**Weather** is good today" is from current project, which not found language key in en-US resource, and it automatically fallback to en resource, it works.
![image](https://github.com/abpframework/abp/assets/25471485/3c751c97-2a13-42bc-93ac-bb6f68c585d3)

```

**Reply by maliming** [reply]
> in my testing, here is the point:

Can you share a simple project to show this case?

**Comment by tian-jie** [comment]
Hi Maliming,

I kept testing and found out the problem.

I made two changes
1. referenced to VFS, I added following code
``` xml
<GenerateEmbeddedFilesManifest>true</GenerateEmbeddedFilesManifest> into the <PropertyGroup>...</PropertyGroup> section of your .csproj file.
```
2. I removed parameters in AddEmbedded
``` csharp
options.FileSets.AddEmbedded<TestWpfModule>();
```
3. **the most import**, I made a different folder name in my solution
``` csharp
            options.Resources
                .Get<LocalizationResources>()
                .AddVirtualJson("/Resources/**AnotherLocation**/")
                .DefaultCultureName = "zh-Hans";
            options.DefaultResourceType = typeof(LocalizationResources);
```

here is the final result which it works fine!
![image](https://github.com/abpframework/abp/assets/25471485/5da5b700-8d35-4ee0-aa7c-2d1b9c6b81b3)

any way, your article and your examples help me to find out the problem. Thank you very much!
